File: frontend/src/types/report.ts
export interface ReportGalleryCategory {
  category: 'signal' | 'background' | 'interesting' | null;
  rank: number | null;
  selection_method: string | null;
}

export interface ReportEventSummary {
  event_id: number;
  features: Record<string, number>;
  source_split: string;
  gallery: ReportGalleryCategory | null;
}

export interface ReportClassification {
  model_id: 'xgboost';
  signal_probability: number;
  predicted_label: 'signal' | 'background';
  threshold: number;
}

export interface FeatureAttribution {
  feature: string;
  value: number;
  contribution: number;
}

export interface ObjectGroupAttribution {
  group: string;
  total_abs_contribution: number;
  signed_contribution: number;
}

export interface ReportExplanation {
  base_value: number;
  margin: number;
  attributions: FeatureAttribution[];
  object_groups: ObjectGroupAttribution[];
}

export interface ReproducibilityModelArtifact {
  model_id: string;
  feature_schema_version: string;
  device: string;
  training_run_origin: string;
  subsample_notes: string;
}

export interface ReproducibilityDataset {
  record: string;
  doi: string;
  content_hash: string;
}

export interface ReproducibilityInferenceContract {
  feature_count: number;
  sentinel_value: number;
  prediction_path: string;
  explanation_path: string;
}

export interface ReproducibilityManifest {
  report_contract_version: string;
  dataset: ReproducibilityDataset;
  inference_contract: ReproducibilityInferenceContract;
  certified_models: ReproducibilityModelArtifact[];
  frozen_leaderboard_status: string;
  server_capabilities: {
    supported_report_formats: string[];
    retraining_supported: boolean;
    public_holdout_access: boolean;
    champion_model_id: string;
  };
}

export interface EventAnalysisReport {
  report_version: string;
  generated_at: string;
  event: ReportEventSummary;
  classification: ReportClassification;
  explanation: ReportExplanation;
  reproducibility: ReproducibilityManifest;
  provenance: {
    statement: string;
  };
}

File: backend/app/schemas/report.py
"""
Pydantic v2 schemas for HiggsLens Research Reports and Reproducibility Manifests.
"""

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

from app.schemas.explain import FeatureAttribution, ObjectGroupAttribution


class ReportGalleryCategory(BaseModel):
    category: Optional[str] = Field(None, description="Gallery category ('signal', 'background', 'interesting') or None")
    rank: Optional[int] = Field(None, description="Gallery ranking if present")
    selection_method: Optional[str] = Field(None, description="Selection method used in gallery")


class ReportEventSummary(BaseModel):
    event_id: int
    features: Dict[str, float]
    source_split: str = "test"
    gallery: Optional[ReportGalleryCategory] = None


class ReportClassification(BaseModel):
    model_id: Literal["xgboost"] = "xgboost"
    signal_probability: float
    predicted_label: Literal["signal", "background"]
    threshold: float


class ReportExplanation(BaseModel):
    base_value: float
    margin: float
    attributions: List[FeatureAttribution]
    object_groups: List[ObjectGroupAttribution]


class ReproducibilityModelArtifact(BaseModel):
    model_id: str = "xgboost"
    feature_schema_version: str = "v1"
    device: str = "CPU (scikit-learn)"
    training_run_origin: str = "Official baseline"
    subsample_notes: str = "Full dataset"


class ReproducibilityDataset(BaseModel):
    record: str = "328"
    doi: str = "10.7483/OPENDATA.ATLAS.ZBP2.M5T8"
    content_hash: str = "54242acf28a78ce303ea48bcf7002f0a44df08448271477e0a63331486c4f316"


class ReproducibilityInferenceContract(BaseModel):
    feature_count: int = 30
    sentinel_value: float = -999.0
    prediction_path: str = "certified PredictionService"
    explanation_path: str = "native XGBoost pred_contribs=True"


class ReproducibilityManifest(BaseModel):
    report_contract_version: str = "1.0"
    dataset: ReproducibilityDataset
    inference_contract: ReproducibilityInferenceContract
    certified_models: List[ReproducibilityModelArtifact]
    frozen_leaderboard_status: str = "Certified & Frozen"
    server_capabilities: Dict[str, Any] = Field(
        default_factory=lambda: {
            "supported_report_formats": ["json", "html"],
            "retraining_supported": False,
            "public_holdout_access": False,
            "champion_model_id": "xgboost"
        }
    )


class EventAnalysisReport(BaseModel):
    report_version: str = "1.0"
    generated_at: str = Field(..., description="Report generation time in ISO-8601 UTC format")
    event: ReportEventSummary
    classification: ReportClassification
    explanation: ReportExplanation
    reproducibility: ReproducibilityManifest
    provenance: Dict[str, str] = Field(
        default_factory=lambda: {
            "statement": "ATLAS open data (record 328, DOI 10.7483/OPENDATA.ATLAS.ZBP2.M5T8) — official ATLAS simulated events, classified by certified pre-trained models."
        }
    )
